sentry-android/src/main/java/io/sentry/android/adapters/UnknownAdapter.java

```diff
@@ -0,0 +1,60 @@
+package io.sentry.android.adapters;
+
+import com.google.gson.*;
+import io.sentry.SentryEvent;
+import io.sentry.UnknownField;
+import io.sentry.protocol.SentryId;
+import java.lang.reflect.Type;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class UnknownAdapter<T extends UnknownField> implements JsonDeserializer<T> {
+
+  private Gson gson =
+      new GsonBuilder()
+          .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
+          .registerTypeAdapter(SentryId.class, new SentryIdSerializerAdapter())
+          .registerTypeAdapter(SentryId.class, new SentryIdDeserializerAdapter())
+          .registerTypeAdapter(Date.class, new DateSerializerAdapter())
+          .registerTypeAdapter(Date.class, new DateDeserializerAdapter())
+          .create();
+
+  @Override
+  public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+      throws JsonParseException {
+    T object = gson.fromJson(json, typeOfT);
+
+    if (object != null && json.isJsonObject()) {
+      JsonObject jsonObject = json.getAsJsonObject();
+
+      Set<Map.Entry<String, JsonElement>> entries = jsonObject.entrySet();
+
+      List<String> keys = getKeys(object);
+
+      if (keys != null && !keys.isEmpty()) {
+        for (Map.Entry<String, JsonElement> item : entries) {
+          if (!keys.contains(item.getKey())) {
+            object.getUnknown().put(item.getKey(), gson.fromJson(item.getValue(), Object.class));
+          }
+        }
+      }
+    }
+
+    return object;
+  }
+
+  private List<String> getKeys(T clazz) {
+    List<String> keys = null;
+
+    if (clazz instanceof SentryEvent) {
+      keys = SentryEvent.KEYS;
```

Comment:
Since we need to be aware of the types here anyway I'd suggest we don't go the inheritance path (i.e forcing the protocol to inherit form `UnknownField` but instead just have a static map:
`Map<Class, Map<String>>`. Here we can get get the right map by the `clazz`.
We can also create the map lazily via reflection on the first call to this adapter to avoid having to keep a manual list of the properties. What do you think?

Reply:
it seems a good idea, but I'd rather have a look at the new PR, this is a crap hack